src/metabase/lib/walk/util.cljc

```diff
@@ -47,3 +48,29 @@
                                 (mapcat :template-tags)
                                 (map (fn [[template-tag-name template-tag]]
                                        (assoc template-tag :lib.walk/template-tag-name template-tag-name))))))
+
+(mu/defn any-native-stage?
+  "Returns true if any stage of this query is native."
+  [query :- ::lib.schema/query]
+  (let [has-native-stage? (volatile! false)]
+    (lib.walk/walk-stages
+     query
+     (fn [_query _path stage]
+       (when (and (not @has-native-stage?)
+                  (= (:lib/type stage) :mbql.stage/native))
+         (vreset! has-native-stage? true))
+       nil))
+    @has-native-stage?))
```

Comment:
An ugly option to short-circuit the search would be to throw an exception when a native stage is found.